### PR TITLE
Avoiding the change of l:binary flag

### LINKED ISSUE #27: Gzipped help files in older version of vim
Hi,

it seems that commit 7a1610f introduced an issue with gzipped help-files (actually gzipped files in general).
According to the commit history, it seems that you tried to fix that with later commits.

However, unfortunately it doesn't work in vim 7.2  (redhat 6.5). E.g. `:help :help` causes 
```
Error detected while processing function gzip#read:
line 44:
Error: Could not read uncompressed file
E434: Can't find tag pattern
```
Checking out revision 2793233, everything works fine.

Regards
Hagen

### PR DESCRIPTION
**Pull request by fidian**
Seems to close #27. Would love others to see if this also solves the problems.

This was inspired by https://github.com/vim/vim/issues/1841#issuecomment-315794374 and the suggestion that we were overwriting `l:binary`.

The only weird bit is that when using `vim -b doc/some-file.txt.gz`, I can never edit it in hexmode. I think this is acceptable.

### CODE REVIEW

**Comment by chaoren** [comment]
Patch looks good to me. Tested it a bit manually, and everything seems fine (though everything *was* fine for me already). 

> The only weird bit is that when using vim -b doc/some-file.txt.gz

This is expected. Because `IsHexmodeEditable` prioritizes checking for gzipped help files over checking for `binary` flag, since manually adding the `binary` flag is indistinguishable for gzipped help files.

>  I can never edit it in hexmode

You can manually use `:Hexmode` after opening the file. But It'd be pretty weird for someone to want to hex edit a temporarily extracted text file.

**Comment by hcgraf** [comment]
`:help` works, and editing files, manually switching to/from Hexmode works.
Only problem here: opening a `.txt.gz` file which is *not* in a `doc` folder (error message from `gzip.vim`, then the compressed file is opened in Hexmode).
Maybe just remove the `doc` from the regex in line 70? It's not really needed I think…

**Comment by fidian** [comment]
@chaoren and @hcgraf - Thanks for testing this patch so quickly!

Ok, I've changed it so the gzip plugin can always win, even if it isn't loaded or doesn't properly edit the file. I'd welcome all feedback and suggestions for improvements.

File: plugin/hexmode.vim [hunk 1]
```diff
@@ -66,8 +66,15 @@ endfunction
...
-    if expand('<afile>:p') =~ '/doc/[^/]*\.txt\.gz$'
+    " The gzip plugin will force the binary flag to be enabled in
+    " BufReadPre, so depending on plugin execution order the flag may or may
+    " not be forced in our BufReadPre but we should always see it in
+    " BufReadPost. We can't watch to see when this flag is flipped because
+    " `vim -b file.gz` starts with it set at the beginning. Instead, we
+    " simply disable Hexmode when editing files the gzip plugin might affect
+    " because there's no way to disable gzip.
+    " See https://github.com/fidian/hexmode/issues/27
+    if expand('<afile>:p') =~ '.*\.\(bz2\|gz\|lzma\|xz\|Z\)$'
```
Comment (chaoren) [inline plugin/hexmode.vim]: The `.*` at the start of the regex is redundant.
Do you need `.zip` here as well?

**Review by chaoren — CHANGES_REQUESTED** [review]

File: plugin/hexmode.vim (hunk 1 above)
Comment (chaoren) [reply plugin/hexmode.vim]: Oh did you lift the regex directly from the gzip plugin? If that's the case then I guess it's better to keep them identical.

**Comment by fidian** [comment]
I've removed the `.*` from the beginning of the pattern. I didn't realize that you had to escape special characters in the pattern, so the `.*` was from some of my early attempts to make it work. Good catch.

The matching on file extensions comes from the gzip plugin. They use this line:

    autocmd BufReadPre,FileReadPre    *.gz,*.bz2,*.Z,*.lzma,*.xz setlocal bin

Because I don't know how to make a `BufReadPre` operate on all extensions *except* the ones that the gzip plugin uses, I thought that this was the cleanest way to test the file extension.

One thing I was pondering is if I should only exclude those file extensions if the gzip plugin is loaded. I decided to not worry about it because it's typically shipped with the system and so almost all installations of Vim have that plugin.

    if exists("loaded_gzip") && ...

The gzip plugin doesn't use .zip. If other plugins conflict with Hexmode, I suppose I will have to exclude them as well. :-/ For instance, I found a plugin that does transparent encryption with GPG and, if it ever shows up here as a bug, I will add another comment and test to ensure we're not editing GPG files if that plugin is loaded.

File: plugin/hexmode.vim [hunk 2]
```diff
@@ -74,7 +74,7 @@ function! s:IsHexmodeEditable()
     " simply disable Hexmode when editing files the gzip plugin might affect
     " because there's no way to disable gzip.
     " See https://github.com/fidian/hexmode/issues/27
-    if expand('<afile>:p') =~ '.*\.\(bz2\|gz\|lzma\|xz\|Z\)$'
```
Comment (chaoren) [inline plugin/hexmode.vim]: Need to keep the escaped dot: `\.` or it'll match anything.

**Comment by chaoren** [comment]
> Because I don't know how to make a BufReadPre operate on all extensions except the ones that the gzip plugin uses

Don't think it's possible.

>  I thought that this was the cleanest way to test the file extension.

Agreed.

> The gzip plugin doesn't use .zip.

Maybe worth looking at the zip plugin and see if it uses `set bin` too.

**Comment by fidian** [comment]
I peeked at those plugins and they seem to do things differently than gzip. I'm not entirely sure how they work, but I tested Hexmode versus a zip and a tar file and they both seemed to work as expected: Hexmode didn't ever kick in even with `vim -b`.

Shortened the comment and fixed the regular expression. I only intended to delete two characters, so it's great that you spotted my mistake.

**Comment by chaoren** [comment]
Looks good. 👍

**Comment by hcgraf** [comment]
Looks good to me too!

**Comment by fidian** [comment]
Thanks for the review!

**Pull request merged**